File: src/main/java/me/halqq/aurora/client/api/util/utils/Colour.java
// Decompiled with CFR 0.152
// Recompile with -Xlint for deprecapted features
package me.halqq.aurora.client.api.util.utils;

import net.minecraft.client.renderer.GlStateManager;

import java.awt.*;

public class Colour {
    public int r, g, b, a;
    public float r1, g1, b1, a1;

    private boolean isInt = true;

    public Colour(int r, int g, int b, int a) {
        this.r = r;
        this.g = g;
        this.b = b;
        this.a = a;
    }

    public Colour(int r, int g, int b) {
        this.r = r;
        this.g = g;
        this.b = b;
        this.a = 255;
    }

    public Colour(Colour color, int a) {
        this.r = color.r;
        this.g = color.g;
        this.b = color.b;
        this.a = a;
    }

    public Colour(float r, float g, float b) {
        this.r1 = r;
        this.g1 = g;
        this.b1 = b;
        isInt = false;
    }

    public Colour(float r, float g, float b, float a) {
        this.r1 = r;
        this.g1 = g;
        this.b1 = b;
        this.a1 = a;
        isInt = false;
    }

    public Colour(int color) {
        this.r1 = ColorUtils.getRed(color);
        this.g1 = ColorUtils.getGreen(color);
        this.b1 = ColorUtils.getBlue(color);
        this.a1 = 1;
        isInt = false;
    }

    public Color getColor() {
        if(isInt) {
            return new Color(r, g, b, a);
        } else {
            return new Color(r1 * 255, g1 * 255, b1 * 255, a1 * 255);
        }
    }

    public int getRGB() {
        if(isInt) {
            return new Color(r, g, b, a).getRGB();
        } else {
            return new Color(r1 * 255, g1 * 255, b1 * 255, a1 * 255).getRGB();
        }
    }

    public float getR() {
        if(isInt) {
            return (float) r / 255;
        } else {
            return r1;
        }
    }

    public float getG() {
        if(isInt) {
            return (float) g / 255;
        } else {
            return g1;
        }
    }

    public float getB() {
        if(isInt) {
            return (float) b / 255;
        } else {
            return b1;
        }
    }

    public float getA() {
        if(isInt) {
            return (float) a / 255;
        } else {
            return a1;
        }
    }

    public int getAlpha() {
        if(isInt) {
            return a;
        } else {
            return (int) a1 * 255;
        }
    }

    public void glColor() {
        if(isInt) {
            GlStateManager.color(r / 255f, g / 255f, b / 255f, a / 255f);
        } else {
            GlStateManager.color(r1, g1, b1, a1);
        }
    }

    public static class ColorUtils {
        public static Color rainbow() {
            long offset = 999999999999L;
            float fade = 1.0f;
            float hue = (float) (System.nanoTime() + offset) / 1.0E10f % 1.0f;
            long color = Long.parseLong(Integer.toHexString(Color.HSBtoRGB(hue, 1.0f, 1.0f)), 16);
            Color c = new Color((int) color);
            return new Color((float) c.getRed() / 255.0f * fade, (float) c.getGreen() / 255.0f * fade, (float) c.getBlue() / 255.0f * fade, (float) c.getAlpha() / 255.0f);
        }

        public static int color(int r, int g, int b, int a) {
            return new Color(r, g, b, a).getRGB();
        }

        public static int color(float r, float g, float b, float a) {
            return new Color(r, g, b, a).getRGB();
        }

        public static int getColor(int a, int r, int g, int b) {
            return a << 24 | r << 16 | g << 8 | b;
        }

        public static int getColor(int r, int g, int b) {
            return 255 << 24 | r << 16 | g << 8 | b;
        }

        public static int astolfoColors(int yOffset, int yTotal) {
            float hue;
            float speed = 2900.0f;
            for (hue = (float) (System.currentTimeMillis() % (long) ((int)speed)) + (float) ((yTotal - yOffset) * 9); hue > speed; hue -= speed) {
            }
            if ((double) (hue /= speed) > 0.5) {
                hue = 0.5f - (hue - 0.5f);
            }
            return Color.HSBtoRGB(hue += 0.5f, 0.5f, 1.0f);
        }

        public static int getRed(int color) {
            return new Color(color).getRed();
        }

        public static int getGreen(int color) {
            return new Color(color).getGreen();
        }

        public static int getBlue(int color) {
            return new Color(color).getBlue();
        }

        public static int getAlpha(int color) {
            return new Color(color).getAlpha();
        }

        public static int rainbow(int delay, long index) {
            double rainbowState = Math.ceil(System.currentTimeMillis() + index + (long)delay) / 15.0;
            return Color.getHSBColor((float)((rainbowState %= 360.0) / 360.0), 0.4f, 1.0f).getRGB();
        }

        public static Color rainbow(final int delay, final float s, final float b) {
            return Color.getHSBColor((System.currentTimeMillis() + delay) % 11520L / 11520.0f, s, b);
        }

        public static int getColor(int brightness) {
            return ColorUtils.getColor(brightness, brightness, brightness, 255);
        }

        public static int getColor(int brightness, int alpha) {
            return ColorUtils.getColor(brightness, brightness, brightness, alpha);
        }

        public static void glColor(final int hex, final int alpha) {
            final float red = (hex >> 16 & 0xFF) / 255F;
            final float green = (hex >> 8 & 0xFF) / 255F;
            final float blue = (hex & 0xFF) / 255F;

            GlStateManager.color(red, green, blue, alpha / 255F);
        }
    }
}

File: src/main/java/me/halqq/aurora/client/api/gui/util/Colour.kt
// Decompiled with CFR 0.152
// Recompile with -Xlint for deprecapted features
package me.halqq.aurora.client.api.gui.util

import net.minecraft.client.renderer.GlStateManager
import net.minecraft.util.math.Vec3d
import java.awt.Color

class Colour {

    @JvmField    var r: Int
    @JvmField    var g: Int
    @JvmField    var b: Int
    @JvmField    var a = 0
    @JvmField   var r1: Float
    @JvmField   var g1: Float
    @JvmField  var b1: Float
    @JvmField   var a1 = 0f
    @JvmField  var isInt = true

    constructor(hsb: FloatArray) : this(Color.getHSBColor(hsb[0], hsb[1], hsb[2])) {}
    constructor(r: Int, g: Int, b: Int, a: Int) {
        this.r = r
        this.g = g
        this.b = b
        this.a = a
        r1 = r / 255f
        g1 = g / 255f
        b1 = b / 255f
        a1 = a / 255f
        fixColorRange()
    }

    constructor(r: Double, g: Double, b: Double, a: Double) : this(r.toInt(), g.toInt(), b.toInt(), a.toInt()) {}
    constructor(r: Int, g: Int, b: Int) {
        this.r = r
        this.g = g
        this.b = b
        a = 255
        r1 = r / 255f
        g1 = g / 255f
        b1 = b / 255f
        a1 = 1f
        fixColorRange()
    }

    constructor(color: Colour, a: Int) {
        r = color.r
        g = color.g
        b = color.b
        this.a = a
        r1 = r / 255f
        g1 = g / 255f
        b1 = b / 255f
        a1 = a / 255f
        fixColorRange()
    }

    constructor(color: Color, a: Float) : this(color) {
        this.a = a.toInt() * 255
        a1 = a
        fixColorRange()
    }

    constructor(r: Float, g: Float, b: Float) {
        r1 = r
        g1 = g
        b1 = b
        this.r = r1.toInt() * 255
        this.g = g1.toInt() * 255
        this.b = b1.toInt() * 255
        isInt = false
        fixColorRange()
    }

    constructor(r: Float, g: Float, b: Float, a: Float) {
        r1 = r
        g1 = g
        b1 = b
        a1 = a
        this.r = r1.toInt() * 255
        this.g = g1.toInt() * 255
        this.b = b1.toInt() * 255
        this.a = a1.toInt() * 255
        isInt = false
        fixColorRange()
    }

    constructor(color: Int) {
        r = ColorUtils.getRed(color)
        g = ColorUtils.getGreen(color)
        b = ColorUtils.getBlue(color)
        a = ColorUtils.getAlpha(color)
        r1 = r / 255f
        g1 = g / 255f
        b1 = b / 255f
        a1 = a / 255f
        fixColorRange()
    }

    fun toVec3d(): Vec3d {
        return Vec3d(r1.toDouble(), g1.toDouble(), b1.toDouble())
    }

    fun nextColor() {
        val hsb = RGBtoHSB()
        var rainbowState = Math.ceil((System.currentTimeMillis() + 200) / 20.0)
        rainbowState %= 360.0
        hsb[0] = (rainbowState / 360.0).toFloat()
        setColour(fromHSB(hsb, a))
    }

    fun syns(color: Colour) {
        r = color.r
        g = color.g
        b = color.b
        a = color.a
        r1 = r / 255f
        g1 = g / 255f
        b1 = b / 255f
        a1 = a / 255f
        fixColorRange()
    }

    fun setBrightness(brightness: Float): Colour {
        val hsb = RGBtoHSB()
        setColour(Colour(floatArrayOf(hsb[0], hsb[1], brightness)))
        return this
    }

    fun setAlpha(alpha: Float): Colour {
        a = (alpha * 255f).toInt()
        a1 = alpha
        fixColorRange()
        return this
    }

    fun setHue(hue: Float): Colour {
        val hsb = RGBtoHSB()
        setColour(Colour(floatArrayOf(hue, hsb[1], hsb[2])))
        return this
    }

    fun setSaturation(saturation: Float): Colour {
        val hsb = RGBtoHSB()
        setColour(Colour(floatArrayOf(hsb[0], saturation, hsb[2])))
        return this
    }

    private fun setColour(color: Colour) {
        r = color.r
        g = color.g
        b = color.b
        a = color.a
        r1 = r / 255f
        g1 = g / 255f
        b1 = b / 255f
        a1 = a / 255f
        fixColorRange()
    }

    val hue: Float
        get() = RGBtoHSB()[0]

    fun RGBtoHSB(): FloatArray {
        return Color.RGBtoHSB(r, g, b, null)
    }

    private fun fixColorRange() {
        if (r > 255) r = 255 else if (r < 0) r = 0
        if (g > 255) g = 255 else if (g < 0) g = 0
        if (b > 255) b = 255 else if (b < 0) b = 0
        if (a > 255) a = 255 else if (a < 0) a = 0
        if (r1 > 1) r1 = 1f else if (r1 < 0) r1 = 0f
        if (g1 > 1) g1 = 1f else if (g1 < 0) g1 = 0f
        if (b1 > 1) b1 = 1f else if (b1 < 0) b1 = 0f
        if (a1 > 1) a1 = 1f else if (a1 < 0) a1 = 0f
    }

    @JvmOverloads
    constructor(color: Color = Color.RED) : this(color.rgb) {
    }

    val color: Color
        get() = Color(r, g, b, a)
    val rGB: Int
        get() = Color(r, g, b, a).rgb

    fun getR(): Float {
        return if (isInt) r.toFloat() / 255 else r1
    }

    fun getG(): Float {
        return if (isInt) g.toFloat() / 255 else g1
    }

    fun getB(): Float {
        return if (isInt) b.toFloat() / 255 else b1
    }

    fun getA(): Float {
        return if (isInt) a.toFloat() / 255 else a1
    }

    val alpha: Int
        get() = if (isInt) a else a1.toInt() * 255

    fun glColor() {
        GlStateManager.color(r1, g1, b1, a1)
    }

    companion object {
        var COLOR_RAINBOW = 0
        var COLOR_ASTOLFO = 0
        var COLOR_PRIMARY = 0

        init {
            COLOR_RAINBOW = 0
            COLOR_ASTOLFO = 1
            COLOR_PRIMARY = 2
        }

        fun fromHSB(hsb: FloatArray, alpha: Int): Colour {
            return Colour(
                ColorUtils.injectAlpha(
                    Color.getHSBColor(
                        hsb[0], hsb[1], hsb[2]
                    ), alpha
                )
            )
        }
    }
}